Global Styles: Fix per block text color customization for cover block

jeyip opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

Partially fixes https://github.com/WordPress/gutenberg/issues/34150

## Description
<!-- Please describe what you have changed or added -->
Block level global styling for cover block text colors were being overridden by default colors. This was because, for global styles, we were setting styles on the top-level cover block element. The default text color, however, was being set on a child, (on the cover block inner container).

Since the text content's nearest ancestor was the inner container, it was inheriting the inner container's text color declarations instead of global styles.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Activate a block theme like `blockbase` or `mayland-blocks`
2. Activate the site editor
3. Navigate to the site editor
4. Insert a cover block onto the page and add text content
5. Open the global styles sidebar, and select the "By Block Type" menu tab
6. Unfold the Cover block-level global styles panel and select different text color options
7. Confirm that the text color options now change according to user customizations

## Screenshots <!-- if applicable -->
![2021-08-19 17 07 56](https://user-images.githubusercontent.com/5414230/130159411-7968c114-161c-447a-b1d3-02a652e9a550.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jeyip committed:
Update specificity of text color declaration

Block level global styling for cover block text colors were being overridden by default colors. This is because, for global styles, we were setting styles on the top-level cover block element. The default text color, however, was being set on the child cover block inner container.

Since the text content's nearest ancestor was the inner container, it was inheriting the inner containers text color declarations.

jeyip reviewed (commented):
**Notes:**

- Alternatively, rather than making the default text color styling less specific like we do in this PR, we could make specificity of the global styles selector _more_ specific. The reason I opted for this approach was because it seemed like it had the smallest surface area of changes to make.
- As far as **background colors** go, there are two places where we can set them on the cover block -- the block settings panel, and the block-level global styling panel. When no selection is made in block settings, background colors chosen in block-level global stylings should take effect. Unfortunately, they don't, and it's another bug that I'm looking into.

creativecoder commented:
> Alternatively, rather than making the default text color styling less specific like we do in this PR, we could make specificity of the global styles selector more specific. The reason I opted for this approach was because it seemed like it had the smallest surface area of changes to make.

I think going with a less specific selector here is a good call; generally keeping selector specificity low for base styles seems like a good practice.

creativecoder reviewed (approved):
✅ This fixes the style conflict for me.

Moving the color declaration to the parent div seems okay--everything inside the inner container seems to inherit the text color without issue.

creativecoder commented:
> I don't see a background color setting for the cover block in the block sidebar settings...

Strike that comment, I forgot you can insert a cover block without an image.

Copons commented:
Tested with TT1-Blocks and it works as expected! ✨ 

---

Although, I've noticed that the Cover's background color doesn't work as well, and I guess we might want to fix that here as well.
Or we could merge this PR right away and open a separate one for the background color, whatever @jeyip prefers. 😄 

When we add the block, we are kinda forced to choose a background color (or an image, but we can ignore that here).
The custom block-level background color correctly overrides any Global Styles customization.

Though, it's possible to _clear_ the block's background color, which becomes hardcoded to black:
https://github.com/WordPress/gutenberg/blob/292d9f2f8752c0aec610113f7f893739e32021e4/packages/block-library/src/cover/style.scss#L46-L48

If I read the associated [issue](https://github.com/WordPress/gutenberg/issues/26545) and [PR](https://github.com/WordPress/gutenberg/pull/26569) correctly, it's related to the Opacity setting (`dimRatio`) that controls the opacity of an overlay color (or gradient) over a background image.

In other words, if the block [has a `dimRatio`](https://github.com/WordPress/gutenberg/blob/292d9f2f8752c0aec610113f7f893739e32021e4/packages/block-library/src/cover/edit.js#L632) making the overlay color visible, we _always_ want an overlay color, even if the user cleared it from the block's settings.

This is good and reasonable UX, but...

The block starts with [`dimRatio: 50` by default](https://github.com/WordPress/gutenberg/blob/292d9f2f8752c0aec610113f7f893739e32021e4/packages/block-library/src/cover/block.json#L30-L33), but the `dimRatio` (Opacity) control is [only displayed if the block has a background image](https://github.com/WordPress/gutenberg/blob/292d9f2f8752c0aec610113f7f893739e32021e4/packages/block-library/src/cover/edit.js#L566-L580).
In short, Cover blocks always start up with `hasBackgroundDim`, even if it doesn't really make sense without a background image.
Therefore, the forced overlay color doesn't make sense without a background image as well.

I think the easy solution would be to just check for the image too [here](https://github.com/WordPress/gutenberg/blob/292d9f2f8752c0aec610113f7f893739e32021e4/packages/block-library/src/cover/edit.js#L632):
```diff
- 'has-background-dim': dimRatio !== 0,
+ url && 'has-background-dim': dimRatio !== 0,
```

In my test, that change fixes the background color issue with Global Styles.
I'll ping @sirreal for a sanity check on my investigation, in the off chance I've overlooked some backward compatibility constraints (e.g. should we expect folks relying on `.has-background-dim` even without background image?).

sirreal commented:
> I'll ping @sirreal for a sanity check on my investigation

👋 Hey @copons! Thanks for the ping. It's unlikely I'll be able to get to this soon. I'll be away for several days and have a lot of other things on my plate. Please don't wait on a reply from me 🙇‍♂️

jeyip commented:
> The block starts with dimRatio: 50 by default, but the dimRatio (Opacity) control is only displayed if the block has a background image.
In short, Cover blocks always start up with hasBackgroundDim, even if it doesn't really make sense without a background image.
Therefore, the forced overlay color doesn't make sense without a background image as well.

> I think the easy solution would be to just check for the image too here:

After looking at the related issues you linked, I think you're right @Copons.

[Inline comments](https://github.com/WordPress/gutenberg/blob/292d9f2f8752c0aec610113f7f893739e32021e4/packages/block-library/src/cover/edit.js#L210-L212) about the purpose of `dimRatio` also support the idea that the forced overlay color without a background image don't make sense. "**If an image and color are set**, dimRatio is used to decide what is used for background darkness checking purposes."

jeyip commented:
> In my test, that change fixes the background color issue with Global Styles.

After testing, I see one potential caveat:
- if a background image is selected 
- and if _**no**_ background color is selected from block settings
- then per block global styles don't work for the cover block.

In practice, the problem seems like an edge case. My intuition wants to say that it's negligible, but I'm curious if others have strong opinions.

Either way, I still think Jacopo spotted a legitimate oversight in the implementation of the `hasBackgroundDim` class, and fixing it is a step in the right direction. I'm unsure if there will be backwards compatibility issues -- at the very least, I don't see references in Gutenberg itself that would cause concern.

![2021-08-23 12 37 27](https://user-images.githubusercontent.com/5414230/130508832-fe13d150-0eef-4c71-8fe5-789487373763.gif)

jeyip committed:
Ensure image url exists before adding background dim class

jeyip commented:
> In my test, that change fixes the background color issue with Global Styles.

Hmm...I also noticed that if no background image and no global styles are applied to the cover block, the background color defaults to white, whereas previously, it defaulted to black. This makes me more concerned about backwards compatibility.

If users have cleared any cover block background color selections, when we implement these changes, their cover blocks will change from black to white.

jeyip committed:
Add default background color logic

jeyip commented:
I added logic that addresses all of the aforementioned issues in c9e7375.

The default, black background color style for the cover block was conflicting with block-level global styles. I introduced a `has-default-background-color` classname that's only applied when no background color has been selected. That way, it doesn't compete with user customized global styles.

Copons reviewed (approved):
Thanks for the fixes @jeyip, this works well for me!

Note that the failing tests are pertinent to these changes.
Feel free to ship this PR once they are sorted out. ✨

oandregal reviewed (changes requested):
As per https://github.com/WordPress/gutenberg/issues/34150#issuecomment-904813412 the fix is actually not allowing the user to edit colors in the global styles sidebar for the cover block.

Copons commented:
> As per [#34150 (comment)](https://github.com/WordPress/gutenberg/issues/34150#issuecomment-904813412) the fix is actually not allowing the user to edit colors in the global styles sidebar for the cover block.

Not super clear what is the issue (but no worries I'll wait for the fix to read more about it 😄 ).
Just, would you please ping @jeyip and me so that we can keep a look on it? 🙇‍♂️

oandregal commented:
Sure! I'll tag you both as reviewers.

jeyip commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/34293
